src/runner.rs

```diff
@@ -3,110 +3,132 @@ use std::ops;
-use std::collections::{HashMap, VecDeque};
-use parity_wasm::elements::{Opcode, BlockType, Local};
+use parity_wasm::elements::Local;
 use {Error, Trap, TrapKind, Signature};
 use module::ModuleRef;
+use memory::MemoryRef;
 use func::{FuncRef, FuncInstance, FuncInstanceInternal};
 use value::{
 	RuntimeValue, FromRuntimeValue, WrapInto, TryTruncateInto, ExtendInto,
 	ArithmeticOps, Integer, Float, LittleEndianConvert, TransmuteInto,
 };
 use host::Externals;
-use common::{DEFAULT_MEMORY_INDEX, DEFAULT_TABLE_INDEX, BlockFrame, BlockFrameType};
-use common::stack::StackWithLimit;
+use common::{DEFAULT_MEMORY_INDEX, DEFAULT_TABLE_INDEX};
 use memory_units::Pages;
 use nan_preserving_float::{F32, F64};
+use isa;
 
 /// Maximum number of entries in value stack.
-pub const DEFAULT_VALUE_STACK_LIMIT: usize = 16384;
-/// Maximum number of entries in frame stack.
-pub const DEFAULT_FRAME_STACK_LIMIT: usize = 16384;
+pub const DEFAULT_VALUE_STACK_LIMIT: usize = (512 * 1024) / ::std::mem::size_of::<RuntimeValue>();
 
-/// Function interpreter.
-pub struct Interpreter<'a, E: Externals + 'a> {
-	externals: &'a mut E,
-}
+// TODO: Make these parameters changeble.
+pub const DEFAULT_CALL_STACK_LIMIT: usize = 16 * 1024;
 
 /// Interpreter action to execute after executing instruction.
 pub enum InstructionOutcome {
 	/// Continue with next instruction.
 	RunNextInstruction,
-	/// Branch to given frame.
-	Branch(usize),
+	/// Branch to an instruction at the given position.
+	Branch(isa::Target),
 	/// Execute function call.
 	ExecuteCall(FuncRef),
-	/// End current frame.
-	End,
 	/// Return from current function block.
-	Return,
+	Return(isa::DropKeep),
 }
 
 /// Function run result.
 enum RunResult {
-	/// Function has returned (optional) value.
-	Return(Option<RuntimeValue>),
+	/// Function has returned.
+	Return,
 	/// Function is calling other function.
 	NestedCall(FuncRef),
 }
 
+/// Function interpreter.
+pub struct Interpreter<'a, E: Externals + 'a> {
+	externals: &'a mut E,
+	value_stack: ValueStack,
+}
+
 impl<'a, E: Externals> Interpreter<'a, E> {
 	pub fn new(externals: &'a mut E) -> Interpreter<'a, E> {
+		let value_stack = ValueStack::with_limit(DEFAULT_VALUE_STACK_LIMIT);
 		Interpreter {
 			externals,
+			value_stack,
 		}
 	}
 
 	pub fn start_execution(&mut self, func: &FuncRef, args: &[RuntimeValue]) -> Result<Option<RuntimeValue>, Trap> {
-		let context = FunctionContext::new(
-			func.clone(),
-			DEFAULT_VALUE_STACK_LIMIT,
-			DEFAULT_FRAME_STACK_LIMIT,
-			func.signature(),
-			args.into_iter().cloned().collect(),
-		);
+		for arg in args {
+			self.value_stack
+				.push(*arg)
+				.map_err(
+					// There is not enough space for pushing initial arguments.
+					// Weird, but bail out anyway.
+					|_| Trap::from(TrapKind::StackOverflow)
+				)?;
+		}
+
+		let initial_frame = FunctionContext::new(func.clone());
 
-		let mut function_stack = VecDeque::new();
-		function_stack.push_back(context);
+		let mut call_stack = Vec::new();
```

Comment:
Can we cache this vec in `self` so that we don't need to realloc every call to `start_execution`? Especially useful because we start empty. We just call `Vec::clear` after `run_interpreter_loop`.

Reply:
That will not make any difference since for now, an `Interpreter` is disposable. That is, every call to `FuncInvoke::invoke` will create a new instance of the `Interpreter` and then drop it after it finishes the execution.

Thankfully, this happens only once when invoking a wasm module exported function, which should be pretty infrequent.